### ISSUE TITLE
repackage the mock catalogs into healpixels

### ISSUE THREAD

**Issue by moustakas** [desihub/desitarget#195]
We discussed on a recent Data call that we should adopt a common data model / organization for the mock catalogs by repackage them into healpixels and ensuring uniformity for a minimum set of columns (ra, dec, redshift, etc).  

I'll do this repackaging with the current set of mocks but for subsequent versions of the mocks we'll ask the "mock-makers" to conform to this data model and data organization.

**Comment by apcooper** [comment]
This is something I was thinking of anyway, but I reckoned it was only worth the effort if Data wanted it. Since that's the case, any future MWS mocks will be organized by healpix rather than bricks.

Is there a preferred directory structure for per-healpix files, as there was for bricks? 

When known, it would handy to have a link here to the data model for the 'minimum' columns.

**Comment by sbailey** [comment]
**Option A**: use the same directory structure as the output spectra:

{group}/{nside}-{pixnum}/{prefix}-{nside}-{pixnum}.*

where group = pixnum // 100.

The spectra use nside=64 (0.84 deg2 / pix) but mocks could consider larger pixels.  The spectro pipeline outputs multiple files per pixel and thus groups them in directories, though this hierarchy is a bit overkill for mocks if they only have one file per pixel.  At the same time, if there isn't any actual problem with it, having a directory symmetry with the output spectra would be conceptually handy.

**Option B**: use a directory scheme adapted for mocks with larger pixels, e.g.

{group}/{prefix}-{nside}-{pixnum}.*

or even just drop the group subdirs if using large pixels, e.g. nside=8 or 16.

Either way, I suggest keeping {nside} and {pixnum} in the filename itself so that every file has a unique name and doesn't depend upon its directory hierarchy to be unique within a versioned mock release.  The {prefix} should differentiate different kinds of mocks.

If using large pixels, the mocks shouldn't be obligated to fill an entire pixel if it only covers part of the DESI footprint.  e.g. the nside=8 pixels that touch the DESI footprint cover 26% more area than the actual footprint, while nside=64 only covers 3% more area.  But if using large pixels and not completely filling them, the mock format should have some standardized way of indicating what fraction of the pixel was filled to make density calculations easier.

A snippet of code that could be useful:
```
#- Use nside=1024 to estimate DESI footprint area
import healpy as hp
import desimodel.footprint
pix1024 = desimodel.footprint.tiles2pix(1024, tiles)
desiarea = len(pix1024) * hp.nside2pixarea(1024, degrees=True)
for nside in [8,16,32,64,128]:
    pix = desimodel.footprint.tiles2pix(nside, tiles)
    area = len(pix) * hp.nside2pixarea(nside, degrees=True)
    print('{:3d}  {:.3f}'.format(nside, area/desiarea))
```
```
  8  1.264
 16  1.133
 32  1.063
 64  1.030
128  1.015
```

**Comment by moustakas** [comment]
@forero I just noticed that you self-assigned this issue.  Are you willing / able to take this over?

**Comment by sbailey** [comment]
Further thoughts:  some mocks already come in different files, in which case requesting that those files be grouped by healpix and follow some directly scheme seems reasonable.  Other mocks come in a single file and it seems somewhat unfortunately to break those up.  We could consider allowing single-file mocks as long as we have a clean way of reading just the subset needed for processing a given patch of sky.  e.g. with fitsio, one could read just the RA and DEC columns, make a filter of which rows are needed, and then read just those rows for the other columns.  I'm not sure what the equivalent recipe is for HDF5.  Or with MPI, a single reader could read and broadcast chunks of the file and individual ranks filter and keep the pieces they need.  Either way, the point is to avoid having N>>1 processes working on different parts of the sky having to read the entire mock catalog into memory, even temporarily.

**Comment by moustakas** [comment]
Note that we already do this optimization, @sbailey.  I still think there will be real benefit to grouping the mocks into healpixels and establishing a common data model.

**Comment by apcooper** [comment]
The MWS main (galaxia) mock v0.0.5 at NERSC is now split into NSIDE=8 healpix, organised by the `{group}/{nside}-{pixnum}/{prefix}-{nside}-{pixnum}.*` scheme.

> the mock format should have some standardized way of indicating what fraction of the pixel was filled to make density calculations easier.

Did we decided on the standard way to do this?

**Comment by geordie666** [comment]
@apcooper, if you call `desimodel.io.load_pixweight` (assuming you have all of the `desimodel/data` files set up via svn) then it will return what fraction of each pixel overlaps the DESI footprint. I _think_ that's what you want...

**Comment by apcooper** [comment]
Finally done. New mocks in:
- `$DESI_MOCKS/mws/galaxia/v0.0.5/`
- `$DESI_MOCKS/mws/galaxia/v0.0.5_superfaint/`

Following the `{nside}/{group}/{pixnum}` layout suggested by @sbailey in an offline discussion with @moustakas.

Each mock file has a header field `FCOVER` storing the result of `desimodel.io.load_pixweight`.

In the case of the `v0.0.5` (non-superfaint) version there are *two* row-matched files in each `pixnum` directory, with different prefixes:
- `mock_allsky_galaxia_desi`: the main mock files, currently all the information required by `desitarget`
- `gaia_mock_allsky_galaxia_desi`: supplementary mock Gaia observables. Right now we're only using these for testing additional gaia-based cuts, so `select_mock_targets` etc. shouldn't need to read these yet.

**Comment by moustakas** [comment]
@apcooper Thanks for these.  I've updated the code to read the latest, repackaged Galaxia mocks (currently still in a branch, but related to #252).

However, one issue is that the filename prefix for the `superfaint` sample changes depending on healpixel.  Specifically, whereas the MWS_MAIN mock files (`v0.0.5`) have a fixed prefix of `mock_allsky_galaxia_desi` the superfaint sample (`0.0.5_superfaint`) uses both `mock_superfaint_allsky_galaxia_desi_b10_cap_south` and `mock_superfaint_allsky_galaxia_desi_b10_cap_north`.  

It's not a big deal -- I can try both prefixes in the code -- but I did want to document this and suggest we use a fixed prefix for all files in the next iteration of the mocks.
